File: mssp/archaeology/008-cpython-logging-warn/src/island_test.py
"""The island test, and the live measurement of CPython's own warn alias.

    python src/island_test.py

Section 4 is why this entry exists: mssp-d-001 asked for a rename under a
second host interface, to find out whether one record shape describes both.
It does not — and section 5 shows exactly which field of the schema breaks.
"""
import io
import inspect
import json
import logging
import pathlib
import re
import sys
import warnings

HERE = pathlib.Path(__file__).parent
sys.path.insert(0, str(HERE))

from DMS import report  # noqa: E402
from SCL import policy  # noqa: E402
from SMS import contract, observe  # noqa: E402
from TMS.emitters import current, legacy  # noqa: E402

RECORD = json.loads((HERE / "FMS" / "architecture.json").read_text(encoding="utf-8"))
ALIAS = RECORD["compatibility_aliases"][0]
OBSERVER = RECORD["observers"][ALIAS["equivalence"]["observer"]]

FAILURES = []


def check(label, ok, detail=""):
    print(f"  {'PASS' if ok else 'FAIL'}  {label}{' - ' + detail if detail else ''}")
    if not ok:
        FAILURES.append(label)


def run_pair():
    before = observe.observe(lambda buf: legacy.emit(buf, "disk almost full", current.emit))
    after = observe.observe(lambda buf: current.emit(buf, "disk almost full"))
    return before, after


print("\n== 1. each emitter is an island")
for name in ("current", "legacy"):
    source = (HERE / "TMS" / "emitters" / f"{name}.py").read_text(encoding="utf-8")
    reaches = re.findall(r"^\s*(?:from|import)\s+(\S+)", source, re.M)
    external = [r for r in reaches if r not in ("warnings",)]
    check(f"emitters/{name} names no sibling", not external, ", ".join(external) or "only the stdlib")
check("the legacy emitter takes its delegate as an argument",
      "delegate" in inspect.signature(legacy.emit).parameters,
      "so the old name calls through without naming the unit it calls")

print("\n== 2. the declaration holds, and for the reason it declared")
before, after = run_pair()
result = contract.check(ALIAS, OBSERVER, before, after,
                        policy.accepts_channel_deltas(), policy.never_differ())
check("the alias holds", result["holds"])
check("output is identical", before["output"] == after["output"], repr(before["output"]))
check("return is identical", before["return"] == after["return"])
check("warnings differ, and the record said they would",
      before["warnings"] != after["warnings"]
      and any(d["observation"] == "warnings" for d in ALIAS["equivalence"]["allowed_deltas"]))
check("and the permission was actually exercised",
      not result["unexercised_permissions"],
      "a permission nothing used would say nothing about whether it was needed")

print("\n== 3. the checks can fail")
strict = contract.check(ALIAS, OBSERVER, before, after,
                        accepts_channel_deltas=False, never_differ=policy.never_differ())
check("a deployment that refuses channel deltas turns the same run into a failure",
      not strict["holds"],
      "same observation, different policy, opposite verdict")
forbidden = contract.check(ALIAS, OBSERVER, before, after,
                           accepts_channel_deltas=True, never_differ=["output", "return", "warnings"])
check("a channel listed as never-differ cannot be waived by allowed_deltas",
      not forbidden["holds"],
      "policy outranks the record for the channels it reserves")
drifted = dict(after)
drifted["output"] = "WARNING:something else\n"
broken = contract.check(ALIAS, OBSERVER, before, drifted,
                        policy.accepts_channel_deltas(), policy.never_differ())
check("an output difference fails, because output is never waivable",
      not broken["holds"],
      [why for ch, ok, why, _ in broken["clauses"] if ch == "output"][0])

print("\n== 4. measured against CPython logging 3.14.5 itself")
check("this interpreter is the examined version", sys.version.split()[0] == "3.14.5",
      sys.version.split()[0])

src = pathlib.Path(inspect.getsourcefile(logging)).read_text(encoding="utf-8")
copies = re.findall(r"def warn\(", src)
check("the alias is hand-written three times", len(copies) == 3,
      "Logger.warn, LoggerAdapter.warn, module-level warn")
stacklevels = set(re.findall(r"DeprecationWarning,\s*(\d+)\)", src))
check("every copy uses the same stacklevel", stacklevels == {"2"}, ", ".join(sorted(stacklevels)))
check("every copy delegates rather than reimplements",
      len(re.findall(r"(?:self\.)?warning\(msg, \*args, \*\*kwargs\)", src)) >= 3,
      "warnings.warn(...) then warning(...)")


def cpython_pair(make_old, make_new):
    def observed(make):
        buf = io.StringIO()
        handler = logging.StreamHandler(buf)
        handler.setFormatter(logging.Formatter("%(levelname)s:%(name)s:%(message)s"))
        log = logging.getLogger("island")
        log.handlers[:] = [handler]
        log.setLevel(logging.DEBUG)
        log.propagate = False
        with warnings.catch_warnings(record=True) as caught:
            warnings.simplefilter("always")
            returned = make(log)
        return {"output": buf.getvalue(),
                "warnings": [(w.category.__name__, str(w.message)) for w in caught],
                "return": repr(returned)}
    return observed(make_old), observed(make_new)


observations = {}
for label, old, new in (
    ("Logger", lambda lg: lg.warn("m %s", 1), lambda lg: lg.warning("m %s", 1)),
    ("LoggerAdapter",
     lambda lg: logging.LoggerAdapter(lg, {}).warn("m %s", 1),
     lambda lg: logging.LoggerAdapter(lg, {}).warning("m %s", 1)),
):
    o, n = cpython_pair(old, new)
    observations[label] = (o, n)
    differing = [c for c in ("output", "warnings", "return") if o[c] != n[c]]
    check(f"{label}: exactly one channel differs", differing == ["warnings"], ", ".join(differing))
    check(f"{label}: and it is the deprecation notice",
          o["warnings"] and o["warnings"][0][0] == "DeprecationWarning" and not n["warnings"],
          o["warnings"][0][1] if o["warnings"] else "none")

# The first version of this line was `check(..., True, ...)` — a hardcoded pass
# in the section measuring whether three hand-written copies had drifted, which
# is the one thing this section exists to find out. Compute it.
logger_delta = {c: observations["Logger"][0][c] != observations["Logger"][1][c]
                for c in ("output", "warnings", "return")}
adapter_delta = {c: observations["LoggerAdapter"][0][c] != observations["LoggerAdapter"][1][c]
                 for c in ("output", "warnings", "return")}
# Metron: the FMS said all three copies were measured on all three channels.
# Two were. The module-level one triggers root basicConfig and was not isolated,
# and the README always said so — the record was the thing overclaiming.
check("the record no longer claims all three were measured behaviourally",
      "TWO measured behaviourally" in RECORD["examined"]["measured"]["do_the_three_copies_agree"],
      "three confirmed statically, two measured")
check("the two hand-written copies produce the same delta shape",
      logger_delta == adapter_delta,
      f"Logger {logger_delta} vs LoggerAdapter {adapter_delta}")
check("and their deprecation messages are worded identically",
      observations["Logger"][0]["warnings"][0][1].replace("method", "X")
      == observations["LoggerAdapter"][0]["warnings"][0][1].replace("method", "X"),
      "the duplication has not drifted — measured, not assumed")

print("\n== 5. the schema from mssp-d-001 cannot express this")
draft_allowed_deltas = ["meta.deprecated", "meta.docs.description"]


def resolve(dotted, obj):
    cursor = obj
    for key in dotted.split("."):
        if not isinstance(cursor, dict) or key not in cursor:
            return None
        cursor = cursor[key]
    return cursor


check("the draft's allowed_deltas are dotted FIELD paths",
      all("." in d for d in draft_allowed_deltas), ", ".join(draft_allowed_deltas))
check("and neither resolves against anything here",
      all(resolve(d, before) is None for d in draft_allowed_deltas),
      "the two callables are indistinguishable as objects")
check("the one permitted difference is an observation, not a field",
      isinstance(ALIAS["equivalence"]["allowed_deltas"][0], dict)
      and "observation" in ALIAS["equivalence"]["allowed_deltas"][0],
      json.dumps(ALIAS["equivalence"]["allowed_deltas"][0], ensure_ascii=False))
# Metron, 2026-08-08: the permit was prose and a RuntimeWarning with the wrong
# message still passed. It is a predicate id now, and an unresolvable one fails
# closed.
from SMS import predicates  # noqa: E402
_bad = dict(ALIAS); _bad["equivalence"] = dict(ALIAS["equivalence"],
    allowed_deltas=[{"observation": "warnings", "predicate": "no-such-predicate"}])
_r = contract.check(_bad, OBSERVER, before, after, True, policy.never_differ())
check("an unresolvable predicate id fails closed", not _r["holds"],
      [t for c, ok, t, _ in _r["clauses"] if c == "warnings"][0])
_wrong = dict(before, warnings=[("RuntimeWarning", "not a deprecation and does not name replacement")])
_r2 = contract.check(ALIAS, OBSERVER, _wrong, after, True, policy.never_differ())
check("Metron's exact counter-case now fails", not _r2["holds"],
      [t for c, ok, t, _ in _r2["clauses"] if c == "warnings"][0][:72])
_unnamed = dict(before, warnings=[("DeprecationWarning", "this is old, stop using it")])
_r3 = contract.check(ALIAS, OBSERVER, _unnamed, after, True, policy.never_differ())
check("a right-class warning that does not name the replacement also fails", not _r3["holds"],
      "the predicate reads the message, not just the channel")
check("so the second host interface amended the schema rather than breaking the idea",
      RECORD["the_amendment"].startswith("allowed_deltas must name OBSERVATIONS"),
      "a field path is one kind of observation, not the only kind")

print("\n== 6. what this entry does not claim")
check("upstream states no sunset, and that is recorded as absent rather than wrong",
      ALIAS["sunset"].startswith("unstated"), ALIAS["sunset"])
check("the observer names its own blind spot",
      "stacklevel" in OBSERVER["what_it_cannot_see"], OBSERVER["what_it_cannot_see"])

print()
if FAILURES:
    print(f"  {len(FAILURES)} check(s) failed: {', '.join(FAILURES)}")
    raise SystemExit(1)
print("  island test passed")

File: mssp/examples/008-compatibility-alias/src/island_test.js
// The island test, and the counter-example mssp-d-001 asked for.
//
//   node src/island_test.js
//
// Section 4 is the one Metron named: a declared alias whose behaviour has
// drifted must make the contract check FAIL, and it must fail because of the
// drift rather than for some other reason that happens to be true at the same
// time.

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { checkAlias } from "./SMS/contract.js";
import { aliasFor, load, record, resolve } from "./SMS/registry.js";
import * as policy from "./SCL/policy.js";

const here = path.dirname(fileURLToPath(import.meta.url));
const failures = [];
const report = (label, ok, detail = "") => {
  console.log(`  ${ok ? "PASS" : "FAIL"}  ${label}${detail ? ` - ${detail}` : ""}`);
  if (!ok) failures.push(label);
};

const FIXTURE = [
  '"use strict";',
  'import { a } from "./a.js";',
  "var legacy = 1;",
  "const x = 2;",
  '"use strict";',
];

const check = (alias, oldRule, newRule) =>
  checkAlias(alias, oldRule, newRule, FIXTURE, policy.currentVersion(),
    policy.majorsBetween, policy.maxWindowVersions());

console.log("\n== 1. every authored rule is an island");
{
  const dir = path.join(here, "TMS", "rules");
  const files = fs.readdirSync(dir).filter((f) => f.endsWith(".js"));
  // Four: two current rules, one legacy alias that drifted, and one unrelated
  // rule that exists so the record is not the only thing in the tree. The
  // number is written down so that adding a file without deciding what it is
  // fails here rather than passing quietly.
  report("there are four authored rules", files.length === 4, files.join(", "));
  for (const file of files) {
    const source = fs.readFileSync(path.join(dir, file), "utf8");
    const reaches = [
      /^\s*import\s[^;]*?\sfrom\s+["']([^"']+)["']/gm,
      /^\s*export\s[^;]*?\sfrom\s+["']([^"']+)["']/gm,
      /^\s*import\s+["']([^"']+)["']/gm,
    ].flatMap((re) => [...source.matchAll(re)].map((m) => m[1]));
    report(`rules/${file} reaches nothing`, reaches.length === 0, reaches.join(", ") || "no specifiers");
  }
  // The legacy shim is the interesting one: it is a compatibility alias and it
  // still does not reference the unit it aliases. It reimplements. That is why
  // the structural rule needs no exception and why the drift was possible.
  const legacy = fs.readFileSync(path.join(dir, "legacy-strict.js"), "utf8");
  report("the legacy alias reimplements rather than references",
    !legacy.includes("strict.js") && legacy.includes("use strict"),
    "no sibling reference, so module 02 is satisfied and says nothing about whether it is still equivalent");
}

console.log("\n== 2. a rename is a row in the record, not a property of a file");
{
  const alias = aliasFor("rules/imports-first");
  report("the old name resolves through FMS", resolve("rules/imports-first").name === "rules/first");
  report("and carries its own lifecycle", alias.valid_from === "2.0.0" && alias.sunset === "3.0.0",
    `${alias.valid_from}..${alias.sunset}`);
  report("a current name resolves to itself", resolve("rules/first").via === null);
  report("there is no authored file for the host-constrained old name",
    !fs.existsSync(path.join(here, "TMS", "rules", "imports-first.js")),
    "the host shim is generated into build/host-shims/, outside TMS");
  report("and the record says where it goes", alias.shim.startsWith("generated"), alias.shim);
}

console.log("\n== 3. a declaration that holds, holds for a stated reason");
{
  const alias = aliasFor("rules/imports-first");
  const base = await load(alias.replacement);
  const shim = { ...base, meta: { ...base.meta, deprecated: true } };
  const result = check(alias, shim, base);
  report("the honest alias passes", result.holds, result.problems.join("; ") || "no problems");
  report("and the only difference is one it declared in advance",
    result.deltas.length === 1 && result.deltas[0].field === "meta.deprecated" && result.deltas[0].allowed,
    JSON.stringify(result.deltas));
}

console.log("\n== 4. the counter-example: a declared alias that has drifted");
{
  const alias = aliasFor("rules/legacy-strict");
  const oldRule = await load("rules/legacy-strict");
  const newRule = await load(alias.replacement);
  const result = check(alias, oldRule, newRule);

  report("the drifted alias FAILS the contract", !result.holds, result.problems.join("; "));
  // Not "it failed". Which reason. Yesterday an island test passed on E0753
  // while claiming to prove something about E0432.
  report("and it fails BECAUSE the findings differ",
    result.problems.includes("findings differ"),
    "not merely because something else about it is also wrong");
  report("the differing finding is named", result.deltas.some((d) =>
    d.field === "findings" && d.old.some((f) => f.includes("var is not permitted"))),
    "the old name objects to var; the replacement does not");
  report("it is ALSO past sunset, and that is reported separately",
    result.problems.some((p) => p.startsWith("past sunset")),
    "two independent failures, not one failure counted twice");

  // The verifier verified: remove the drift and the findings clause must stop
  // complaining. If it still complains, the check was reporting on something
  // other than behaviour.
  const faithful = { ...newRule, meta: { ...newRule.meta, deprecated: true } };
  const repaired = check(alias, faithful, newRule);
  report("a faithful shim stops the findings complaint",
    !repaired.problems.includes("findings differ"),
    repaired.problems.join("; ") || "no problems at all");
  report("so the findings clause is what detected the drift",
    result.problems.includes("findings differ") && !repaired.problems.includes("findings differ"));
  report("while the sunset complaint survives the repair",
    repaired.problems.some((p) => p.startsWith("past sunset")),
    "lifecycle and equivalence are independent, and the report keeps them apart");
}

console.log("\n== 5. the contract cannot be satisfied by declaring the drift away");
{
  // The obvious cheat: add findings to allowed_deltas. The observer must refuse
  // it, because an equivalence contract that permits behaviour to differ is not
  // a contract.
  const alias = { ...aliasFor("rules/legacy-strict"),
    equivalence: { observer: "rule-contract-v1", allowed_deltas: ["findings", "meta.deprecated"] } };
  const oldRule = await load("rules/legacy-strict");
  const newRule = await load(alias.replacement);
  const result = check(alias, oldRule, newRule);
  report("listing findings as an allowed delta does not make the alias hold",
    !result.holds, result.problems.join("; "));
  // It now fails EARLIER than it used to. Before 2026-08-08 the refusal came out
  // of the comparison ("findings differ"); the observer now declares findings
  // non-waivable and the contract is rejected at declaration time, before
  // anything is compared. Refusing an unacceptable declaration beats refusing
  // its result, because the result depends on a fixture and the declaration
  // does not.
  report("and it is refused at declaration time, not by the comparison",
    result.problems.some((p) => p.includes("non-waivable")),
    "the observer declares findings non-waivable, so the contract never gets to run");

  // Metron ran this file on 2026-08-08 with an observer id that does not exist
  // and got holds:true — the name was a label and the comparator was hard-coded.
  const bogus = { ...aliasFor("rules/imports-first"),
    equivalence: { observer: "observer-that-does-not-exist", allowed_deltas: ["meta.deprecated"] } };
  const base = await load("rules/first");
  const bogusResult = check(bogus, { ...base, meta: { ...base.meta, deprecated: true } }, base);
  report("an observer id that does not resolve fails closed",
    !bogusResult.holds && bogusResult.problems.some((p) => p.includes("does not resolve")),
    bogusResult.problems.join("; "));

  const offObserver = { ...aliasFor("rules/imports-first"),
    equivalence: { observer: "rule-contract-v1", allowed_deltas: ["timing.ms"] } };
  const offResult = check(offObserver, { ...base, meta: { ...base.meta, deprecated: true } }, base);
  report("an allowed_delta the observer does not observe is refused",
    !offResult.holds, offResult.problems.join("; "));
}

console.log("\n== 6. SCL owns the window, not the unit");
{
  report("only a maintainer may open a compatibility window",
    policy.mayOpenAlias("maintainer") && !policy.mayOpenAlias("contributor"));
  report("the permitted window is one major version", policy.maxWindowVersions() === 1);
  const wide = { ...aliasFor("rules/imports-first"), valid_from: "1.0.0", sunset: "9.0.0" };
  const base = await load("rules/first");
  const result = check(wide, { ...base, meta: { ...base.meta, deprecated: true } }, base);
  report("a window wider than policy allows is refused",
    result.problems.some((p) => p.includes("exceeds the permitted")),
    result.problems.join("; "));
}

console.log("\n== 7. what this run does not claim");
{
  report("the record marks itself as a candidate, not adopted method",
    record.status === "candidate", record.note_on_status.slice(0, 60) + "…");
  report("and names the observer's blind spot",
    record.observers["rule-contract-v1"].what_it_cannot_see.includes("timing"),
    record.observers["rule-contract-v1"].what_it_cannot_see);
}

console.log("");
if (failures.length) {
  console.log(`  ${failures.length} check(s) failed: ${failures.join(", ")}`);
  process.exit(1);
}
console.log("  island test passed");
